templates/cis2-nft/src/lib.rs

```diff
@@ -0,0 +1,1061 @@
+            return Ok(());
+        }
+        // Since this contract only contains NFTs, no one will have an amount greater
+        // than 1. And since the amount cannot be the zero at this point, the
+        // address must have insufficient funds for any amount other than 1.
+        ensure_eq!(amount, 1.into(), ContractError::InsufficientFunds);
+
+        {
+            let mut from_address_state =
+                self.state.get_mut(from).ok_or(ContractError::InsufficientFunds)?;
+            // Find and remove the token from the owner, if nothing is removed, we know the
+            // address did not own the token..
+            let from_had_the_token = from_address_state.owned_tokens.remove(token_id);
+            ensure!(from_had_the_token, ContractError::InsufficientFunds);
+        }
+
+        // Add the token to the new owner.
+        let mut to_address_state =
+            self.state.entry(*to).or_insert_with(|| AddressState::empty(state_builder));
+        to_address_state.owned_tokens.insert(*token_id);
+        Ok(())
+    }
+
+    /// Update the state adding a new operator for a given address.
+    /// Succeeds even if the `operator` is already an operator for the
+    /// `address`.
+    fn add_operator(
+        &mut self,
+        owner: &Address,
+        operator: &Address,
+        state_builder: &mut StateBuilder<S>,
+    ) {
+        let mut owner_state =
+            self.state.entry(*owner).or_insert_with(|| AddressState::empty(state_builder));
+        owner_state.operators.insert(*operator);
+    }
+
+    /// Update the state removing an operator for a given address.
+    /// Succeeds even if the `operator` is _not_ an operator for the `address`.
+    fn remove_operator(&mut self, owner: &Address, operator: &Address) {
+        self.state.entry(*owner).and_modify(|address_state| {
+            address_state.operators.remove(operator);
+        });
+    }
+
+    /// Check if state contains any implementors for a given standard.
+    fn have_implementors(&self, std_id: &StandardIdentifierOwned) -> SupportResult {
+        if let Some(addresses) = self.implementors.get(std_id) {
+            SupportResult::SupportBy(addresses.to_vec())
+        } else {
+            SupportResult::NoSupport
+        }
+    }
+
+    /// Set implementors for a given standard.
+    fn set_implementors(
+        &mut self,
+        std_id: StandardIdentifierOwned,
+        implementors: Vec<ContractAddress>,
+    ) {
+        self.implementors.insert(std_id, implementors);
+    }
+}
+
+/// Build a string from TOKEN_METADATA_BASE_URL appended with the token ID
+/// encoded as hex.
+fn build_token_metadata_url(token_id: &ContractTokenId) -> String {
+    let mut token_metadata_url = String::from(TOKEN_METADATA_BASE_URL);
+    token_metadata_url.push_str(&token_id.to_string());
+    token_metadata_url
+}
+
+// Contract functions
+
+/// Initialize contract instance with no token types initially.
+#[init(contract = "CIS2-NFT")]
```

Comment:
Do we want to keep this name. We did briefly discuss that the name of the contract would be the same as the name of the project, did we not?